gips/0081-indexing-payments.md

```diff
@@ -0,0 +1,165 @@
+
+1. Desired performance should be defined by the payer.
+1. Performance can only be assessed after the service has been consumed.
+1. The cost of serving a subgraph is unknown a-priori, and it depends on the desired query performance as well as the indexing cost.
+
+To solve for the first two, we propose that the gateway is the one coordinating with the indexers. Gateways can define and assess Quality of Service (QoS) on an ongoing basis and decide which indexers to continue to work with based on that.
+
+To solve for the last concern, we propose that indexers are paid proportionally to the amount of work they do to index the subgraph. This amount of work is reported by indexers and thus needs to be sufficiently verifiable for gateways, fishermen or other indexers to dispute how much work was actually done.
+
+In particular we propose that the coordination mechanism is underpinned by an **Indexing Agreement** to be drafted by the gateway, shared with a subset of indexers of its choosing and subsequently approved by some of those indexers. These agreements are designed to allow for indexers to work and collect for their work without having to go back to the payer. They are self-contained in that regard. Furthermore, they specify the following:
+
+1. the subgraph to be indexed
+1. the data service that will initiate payment collection
+1. a deadline by which the agreement must be accepted
+1. an end date to the agreement
+1. a maximum amount payable for the initial indexing.
+1. a maximum amount payable for the ongoing indexing. The ongoing indexing is the indexing that happens between the last collection and the present.
+1. the maximum amount of time that can elapse between collections.
+1. a price per unit of work done.
+
+This agreement is to be upheld by protocol smart contracts, leveraging horizon primitives. This means that payers will have to escrow funds similarly to what's done for query fees. Between the smart contracts and the escrow, indexers are guaranteed that if they accept an indexing agreement, they'll be compensated as agreed, without the need to trust the agreement's counterpart.
+
+At the same time, agreements are cancellable by either party at any time, with some limitations put in place to allow indexers to collect for all work done up to that point.
+
+Putting all of this together the solution consists of the following:
+
+- Indexers disclose publicly their price per unit of work.
+- Gateways send indexing agreements to indexers they want to work with.
+- Indexers accept (on-chain) agreements that meet their criteria and get to work indexing.
+- On a regular basis:
+  - Indexers collect for their work, by posting on-chain the original agreement, a Proof of Indexing (PoI) and the amount of work done for that period.
+  - Gateways monitor the QoS per open agreement and assess if they need to cancel (on-chain) and/or replace any of them.
```

Comment:
I think if 'assess if it meets their requirements and the agreement should be continued, or cancelled if not'? Some duplication between cancel and/or replace to be resolved here.

Reply:
done